dlib/image_loader/jxl_loader.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (C) 2024  Davis E. King (davis@dlib.net), Adrià Arrufat
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_JXL_IMPORT
+#define DLIB_JXL_IMPORT
+
+#include <vector>
+
+#include "jxl_loader_abstract.h"
+#include "image_loader.h"
+#include "../pixel.h"
+#include "../dir_nav.h"
+#include "../test_for_odr_violations.h"
+
+namespace dlib
+{
+
+    class jxl_loader : noncopyable
+    {
+    public:
+
+        jxl_loader(const char* filename);
+        jxl_loader(const std::string& filename);
+        jxl_loader(const dlib::file& f);
+        jxl_loader(const unsigned char* imgbuffer, size_t buffersize);
+
+        bool is_gray() const;
+        bool is_graya() const;
+        bool is_rgb() const;
+        bool is_rgba() const;
+        unsigned int bit_depth() const;
+        unsigned int nr() const;
+        unsigned int nc() const;
+
+        template<typename image_type>
+        void get_image(image_type& image) const
+        {
+#ifndef DLIB_JXL_SUPPORT
+            /* !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+                You are getting this error because you are trying to use the jxl_loader
+                object but you haven't defined DLIB_JXL_SUPPORT.  You must do so to use
+                this object.   You must also make sure you set your build environment
+                to link against the libjxl library.
+            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*/
+            static_assert(sizeof(image_type) == 0, "JPEG XL support not enabled.");
+#endif
+            image_view<image_type> vimg(image);
+            vimg.set_size(height, width);
+            using pixel_type = typename image_traits<image_type>::pixel_type;
+
+            // Fast path: rgb, rgb_alpha, grayscale with matching input depth
+            if (pixel_traits<pixel_type>::num == depth && (
+                pixel_traits<pixel_type>::rgb ||
+                pixel_traits<pixel_type>::rgb_alpha ||
+                pixel_traits<pixel_type>::grayscale))
```

Comment:
This needs to check if it's grayscale and of base pixel type unsigned char too doesn't it?

Reply:
Right, I was thinking that even if it's not an unsigned char, it would work, since JPEG XL supports ints and floats. But I should also tell the encoder the type. So let's just restrain ourselves to unsigned char in this PR.